src/service/vaults.ts

```diff
@@ -92,6 +91,78 @@ const watchPriceFeeds = () => {
   };
 };
 
+type VaultSubscribers = {
+  asset: string;
+  vault: string;
+};
+
+const watchUserVaults = () => {
+  let isStopped = false;
+  const watchedVaults = new Set<string>();
+
+  const watchVault = async (
+    offerId: string,
+    subscriber: string,
+    managerId: string,
+  ) => {
+    useVaultStore.getState().markVaultForLoading(offerId, managerId);
+
+    const { leader, importContext } = appStore.getState();
+    const f = makeFollower(`:${subscriber}`, leader, {
+      unserializer: importContext.fromBoard,
+    });
+
+    for await (const { value } of iterateLatest<VaultUpdate>(f)) {
+      if (isStopped) break;
+      console.debug('got update', subscriber, value);
+      useVaultStore
+        .getState()
+        .setVault(offerId, { ...value, managerId, isLoading: false });
+    }
+  };
+
+  const watchNewVaults = async (
+    subscribers: Record<string, Record<string, string>>,
+  ) => {
+    if (!useVaultStore.getState().vaults) {
+      useVaultStore.setState({ vaults: new Map() });
+    }
+
+    Object.entries(subscribers).forEach(([offerId, subscribers]) => {
+      // XXX: If a third party contract returns a similar looking offer result,
```

Comment:
good point! any offer result can have a `vault` key. to know it's really a vault we have to watch that key under `published.vaultFactory` which is authoritative.

Reply:
It could still get a fake vault path (in that case it'd return no data and we'd show a vault summary card with an error state), or a vault path for a vault you don't own (that could be malicious and trick you into sending a bogus offer). I think it's worth revisiting adding boardIds to these publicSubscriber objects. Even if they can't always have boardIds, some offers (like opening vaults) are created with boardIds specified, and it would be useful to track those back to their offer results.

Reply:
I might be forgetting the implementation details though, I guess the smart wallet derives the path string, and the contract returns the actual reference to the subscriber? In that case we could trust that the path isn't spoofable?

Reply:
Hmm, unless another contract could get ahold of the subscriber since it is public...

Reply:
> I might be forgetting the implementation details though, I guess the smart wallet derives the path string, and the contract returns the actual reference to the subscriber? In that case we could trust that the path isn't spoofable?

Good to think through this. The smart wallet gets the path string from the object, but it's not branded so there's no way to verify what object is returning the string. StorageNodes can only be made under a parent path, but one could make an object that acts like a StorageNode but returns an arbitrary string.

Fortunately we don't have any risk of this until MN-3.

> I think it's worth revisiting adding boardIds to these publicSubscriber objects.

Before Vaults Release?

Reply:
Right no risk until MN-3, so no work to be done at the moment.